XenAdmin/Alerts/Types/ClientUpdateAlert.cs

```diff
@@ -134,17 +136,66 @@ public static void DownloadAndInstallNewClient(ClientUpdateAlert updateAlert, IW
             {
                 Process.Start(outputPathAndFileName);
                 log.DebugFormat("Update {0} found and install started", updateAlert.Name);
-                downloadAndInstallClientAction.ReleaseInstaller();
+                downloadAndInstallClientAction.ReleaseDownloadedContent();
                 Application.Exit();
             }
             catch (Exception e)
             {
                 log.Error("Exception occurred when starting the installation process.", e);
-                downloadAndInstallClientAction.ReleaseInstaller(true);
+                downloadAndInstallClientAction.ReleaseDownloadedContent(true);
 
                 using (var dlg = new ErrorDialog(Messages.UPDATE_CLIENT_FAILED_INSTALLER_LAUNCH))
                     dlg.ShowDialog(parent);
             }
         }
+
+        public static void DownloadSource(IWin32Window parent)
+        {
+            var outputPathAndFileName = string.Empty;
+            using (var saveSourceDialog = new SaveFileDialog())
+            {
+                saveSourceDialog.FileName = BrandManager.BrandConsoleNoSpace + "-source.zip";
+                saveSourceDialog.DefaultExt = "zip";
+                saveSourceDialog.Filter = "(*.zip)|*.zip|All files (*.*)|*.*";
+                saveSourceDialog.InitialDirectory = Win32.GetKnownFolderPath(Win32.KnownFolders.Downloads);
+
+                if (saveSourceDialog.ShowDialog() != DialogResult.OK)
+                {
+                    return;
+                }
+                outputPathAndFileName = saveSourceDialog.FileName;
+            }
+
+            Console.WriteLine(Updates.ClientVersions.Count);
+            var clientVersion = Updates.ClientVersions.First();

```

Comment:
probably meant `FirstOrDefault` here?

> The [First<TSource>(IEnumerable<TSource>)](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.first?view=netframework-4.8.1#system-linq-enumerable-first-1(system-collections-generic-ienumerable((-0)))) method throws an exception if source contains no elements. To instead return a default value when the source sequence is empty, use the [FirstOrDefault](https://learn.microsoft.com/en-us/dotnet/api/system.linq.enumerable.firstordefault?view=netframework-4.8.1) method.

```suggestion
            var clientVersion = Updates.ClientVersions.FirstOrDefault();
```